lib/jack_tokenizer.rb

```diff
@@ -53,11 +53,7 @@ def advance
       @token_type = :IDENTIFIER
     end
 
-    if next_index.nil?
```

Comment:
OK, we got rid of the `next_index = @input.index(" ", @index)` line which could make `next_index` be `nil`, but that’s not the only way it can end up `nil`. What else can cause this?

Reply:
I think `next_index` can be nil from this line `next_index = @input.index('"', @index + 1) + 1`, which is inside the check to see if the current character is an opening quote. It can occur if we pass a string without a closing quote. 

Was this what you were getting at?

Reply:
Yeah, that’s exactly right! If we read the beginning of a string, but then can’t find the end of the string, that’ll put the tokenizer into a bad state. But I don’t think the chapter 10 projects actually ask us to parse any programs with that problem, so it’s up to us whether we want to handle it somehow.

Reply:
(I should clarify that I think `next_index = @input.index('"', @index + 1) + 1` will actually raise an exception in this case, rather than assigning `nil` to `next_index`, because the `nil + 1` part won’t work. So I guess Ruby will already raise an exception for us if we try to parse a malformed string constant, even if the error is a confusing one.)

Reply:
Right! I assumed that we don't need to do any syntax error reporting (at least not yet). Do you think we should try to handle it?

Reply:
Nah, I think it’s fine for now, I just wanted us to think it through and be aware of what else could go wrong. If we do ever have a syntax error we’ll be ready to understand what happened.